### PR TITLE
Restore multiplexed JSONRPC command support

### PR DESCRIPTION
**Pull request by rustyrussell**
(On top of #2081)

This also opens the door to notifications, AFAICT.

### CODE REVIEW

File: tests/test_misc.py [hunk 1]
```diff
@@ -617,6 +617,37 @@ def test_multirpc(node_factory):
...
+    ]
+
+    sock.sendall(b'\n'.join(commands))
+
+    buff = b''
+
+    # They will return in the same order, since they start immediately
+    # (delaying completion should mean we don't see the other commands intermingled).
+    for i in commands:
+        obj, buff = l1.rpc._readobj(sock, buff)
```
Comment (cdecker) [inline tests/test_misc.py]: Probably should also assert that `buff` is empty here.

File: tests/test_misc.py [hunk 2]
```diff
@@ -648,6 +648,61 @@ def test_multiplexed_rpc(node_factory):
...
+        b'{"id":1,"jsonrpc":"2.0","params":[2000]}',
+        # Bad ID
+        b'{"id":{2},"jsonrpc":"2.0","method":"dev-slowcmd","params":[1500]}',
+        b'{"id":2,"jsonrpc":"2.0","method":"dev-slowcmd","params":[1500]}',
+        b'{"id":3,"jsonrpc":"2.0","method":"dev-slowcmd","params":[1000]}',
+        b'{"id":3,"jsonrpc":"2.0","method":"dev-slowcmd","params":[1000]}',
+        b'{"id":4,"jsonrpc":"2.0","method":"dev-slowcmd","params":[500]}',
+        b'{"id":4,"jsonrpc":"2.0","method":"dev-slowcmd","params":[500]}'
+    ]
+    # no ID
```
Comment (cdecker) [inline tests/test_misc.py]: The comments seem to be misaligned with the actual tests.

Comment (cdecker) [reply tests/test_misc.py]: This looks like a good usecase for table driven testing: create a dict with the inputs as keys and the expected outcome as the values and then iterate through them.

**Comment by cdecker** [comment]
Added 3 tiny fixups to see if I can make Travis happy :-)

File: lightningd/jsonrpc.c [hunk 3]
```diff
@@ -357,27 +290,34 @@ static void destroy_command(struct command *cmd)
...
-	if (cmd->jcon)
-		jcon_append(cmd->jcon, " }\n\n");
+	json_stream_append(result, " }\n\n");
+	json_stream_close(result, cmd);
 	if (cmd->ok)
 		*(cmd->ok) = true;
+
+	/* If we have a jcon, it will free result for us. */
+	if (cmd->jcon)
+		tal_steal(cmd->jcon, result);
```
Comment (ZmnSCPxj) [inline lightningd/jsonrpc.c]: If we do not have a `jcon`, does `result` get taken by this function?

Comment (rustyrussell) [reply lightningd/jsonrpc.c]: This function frees `cmd` a few lines below, so `result` goes with it...

File: lightningd/json_stream.c [hunk 4]
```diff
@@ -0,0 +1,289 @@
+#include <ccan/io/io.h>
+  /* To reach into io_plan: not a public header! */
+  #include <ccan/io/backend.h>
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply lightningd/json_stream.c]: Yes, it's ugly.  Removing jsonrpc.c one now...

File: CHANGELOG.md [hunk 5]
```diff
@@ -8,6 +8,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ### Added
 
+- JSON API: use `\n\n` to terminate responses, for simplified parsing.
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply CHANGELOG.md]: Yep, will do.

**Comment by rustyrussell** [comment]
OK, rebased on master, folded @cdecker fixes.  New commit at the head (CHANGELOG tweak) and tail (more documentation).

**Comment by ZmnSCPxj** [comment]
ACK 93f989f

**Comment by cdecker** [comment]
ACK https://github.com/ElementsProject/lightning/pull/2092/commits/93f989f1a14febd095770109d5be6f821debec05

**Pull request merged**